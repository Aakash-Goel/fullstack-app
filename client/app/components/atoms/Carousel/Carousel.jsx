import React from 'react';
import { string, any, object, shape, oneOf } from 'prop-types';
import classnames from 'classnames';
import Slider from 'react-slick';

import withStyles from '@material-ui/core/styles/withStyles';
import { ChevronLeftOutlined, ChevronRightOutlined } from '@material-ui/icons';

import carouselStyle from './Carousel.style';

/**
 * @param  {bool}   accessibility Enables tabbing and arrow key navigation. Defaults to `true`
 * @param  {bool}   adaptiveHeight Adjust the slide's height automatically. Defaults to `false`
 * @param  {func}   afterChange Callback function called after the current index changes. The new index is accessible in the first parameter of the callback function
 * @param  {bool}   arrows Should we show Left and right nav arrows. Defaults to `true`
 * @param  {string} arrowClass Class applied to arrow if they are enabled. Defaults to `''`
 * @param  {bool}   autoplay  Should it be auto scroll. Defaults to `false`
 * @param  {int}    autoplaySpeed Delay between each auto scroll (in ms). Defaults to `3000`
 * @param  {func}   beforeChange  Callback function called before the current index changes. The old and the new indexes are accessible in the first and the second parameters of the callback function respectively
 * @param  {bool}   centerMode  Should we centre to a single item. Defaults to `false`
 * @param  {string} centerPadding Padding for the slide in center. Defaults to `50px`
 * @param  {any}    children Child components or elements that the carousel accepts. This is a required prop
 * @param  {string} className CSS class for inner slider div. Defaults to `''`
 * @param  {func}   customPaging  Custom paging templates. Defaults to `i => <button>{i + 1}</button>`
 * @param  {bool}   dots  Should we show the dots at the bottom of the gallery. Defaults to `false`
 * @param  {string} dotsClass Class applied to the dots if they are enabled. Defaults to `slick-dots`
 * @param  {bool}   draggable Is the gallery scroll via dragging on desktop. Defaults to `true`
 * @param  {string} easing Behavior of scroll elements. Defaults to `linear`
 * @param  {bool}   fade  Slides use fade for transition. Defaults to `false`
 * @param  {bool}   focusOnSelect Go to slide on click. Defaults to `false`
 * @param  {bool}   infinite  Should the gallery infinitely wrap around its contents. Defaults to `true`
 * @param  {bool}   initialSlide  Which item should be the first to be displayed. Defaults to `0`
 * @param  {ondemand`/`progressive}   lazyLoad  Loads images or renders components on demand or progressively. Defaults to `null`
 * @param  {func}   onInit  componentWillMount callback. `() => void`. Defaults to `null`
 * @param  {func}   onLazyLoad  Callback after slides load lazily. `slidesLoaded => {...}`. Defaults to `null`
 * @param  {func}   onReInit  componentDidUpdate callback. `() => void`. Defaults to `null`
 * @param  {func}   onSwipe  Callback after slide changes by swiping. Defaults to `null`
 * @param  {bool}   pauseOnDotsHover  Prevents autoplay while hovering on dots. Defaults to `false`
 * @param  {bool}   pauseOnFocus  Prevents autoplay while focused on slides. Defaults to `false`
 * @param  {bool}   pauseOnHover  Prevents autoplay while hovering. Defaults to `true`
 * @param  {array}  responsive  Array of objects in the form of { breakpoint: int, settings: { ... } } The breakpoint int is the maxWidth so the settings will be applied when resolution is below this value. Breakpoints in the array should be ordered from smallest to greatest. Use 'unslick' in place of the settings object to disable rendering the carousel at that breakpoint. Example:  { breakpoint: 768, settings: { slidesToShow: 3 } }, { breakpoint: 1024, settings: { slidesToShow: 5 } }, { breakpoint: 100000, settings: 'unslick' }
 * @param  {int}    rows  Number of rows per slide in the slider, (enables grid mode). Defaults to `1`
 * @param  {bool}   rtl Reverses the slide order. Defaults to `false`
 * @param  {string} slide Slide container type. Defaults to `div`
 * @param  {int}    slidesPerRow  Number of slides to display in grid mode, this is useful with rows option. Defaults to `1`
 * @param  {int}    slidesToScroll  Number of slides to scroll for each navigation item. Defaults to `1`
 * @param  {int}    slidesToShow  Number of slides to be visible at a time. Defaults to `1`
 * @param  {int}    speed Animation speed in milliseconds. Defaults to `500`
 * @param  {bool}   swipe Enable swipe to change slides. Defaults to `true`
 * @param  {bool}   swipeToSlide  Allow users to drag or swipe directly to a slide irrespective of slidesToScroll. Defaults to `false`
 * @param  {bool}   touchMove Defaults to `true`
 * @param  {int}    touchThreshold Defaults to `5`
 * @param  {bool}   useCSS  Enable/Disable CSS Transitions. Defaults to `true`
 * @param  {bool}   useTransform  Enable/Disable CSS transforms. Defaults to `true`
 * @param  {bool}   variableWidth Defaults to `false`
 * @param  {bool}   vertical  Vertical slide mode. Defaults to `false`
 * @param  {React Element} nextArrow Allows to select a node or customize the HTML for the "Next" arrow. Defaults to `<button type="button" class="slick-next">Next</button>`
 * @param  {React Element} prevArrow Allows to select a node or customize the HTML for the "Previous" arrow. Defaults to `<button type="button" class="slick-prev">Previous</button>`
 * @param  {string} nextArrowClass Class applied to the next arrow only if arrows are enabled. Defaults to `''`
 * @param  {string} prevArrowClass Class applied to the previous arrow only if arrows are enabled. Defaults to `''`
 *
 * @return {html tag} <div> [return a container which represent the carousel
 */
const propTypes = {
  classes: object.isRequired,
  children: any.isRequired,
  customizeArrow: shape({
    color: oneOf([
      'inherit',
      'primary',
      'secondary',
      'action',
      'error',
      'disabled',
    ]),
    fontSize: oneOf(['inherit', 'default', 'small', 'large']),
    arrowClass: string,
    nextArrowClass: string,
    prevArrowClass: string,
  }),
};

const defaultProps = {
  customizeArrow: {
    color: 'primary',
    fontSize: 'default',
    arrowClass: '',
    nextArrowClass: '',
    prevArrowClass: '',
  },
};

/* eslint-disable react/prop-types */
/**
 * SlickPrevButton
 * extract `currentSlide` and `slideCount` from props to avoid passing these props to custom arrow slick container.
 * The reason of this is because these props are giving warning that `React does not recognize the slideCount and currentSlide prop on DOM element`.
 *
 * Fix is being taken from
 *  https://github.com/akiran/react-slick/issues/1195
 *
 */
const SlickPrevButton = ({ currentSlide, slideCount, children, ...props }) => (
  <div alt="alt-prevArrowButton" title="alt-prevArrowButton" {...props}>
    {children}
  </div>
);

/**
 * SlickNextButton
 * extract `currentSlide` and `slideCount` from props to avoid passing these props to custom arrow slick container.
 * The reason of this is because these props are giving warning that `React does not recognize the slideCount and currentSlide prop on DOM element`.
 *
 * Fix is being taken from
 *  https://github.com/akiran/react-slick/issues/1195
 *
 */
const SlickNextButton = ({ currentSlide, slideCount, children, ...props }) => (
  <div alt="alt-nextArrowButton" title="alt-nextArrowButton" {...props}>
    {children}
  </div>
);
/* eslint-enable react/prop-types */

/**
 * This carousel component is implementation of a third party library react-slick.
 * This document is copied from the original documentation of react-slick from github.
 * Please refer to that documentation for further information.
 *
 * @see Slider
 *
 */
class Carousel extends React.PureComponent {
  constructor(props) {
    super(props);
    this.setRef = this.setRef.bind(this);
  }

  setRef(ref) {
    this.slick = ref;
  }

  /**
   * @function render
   */
  render() {
    const { children, customizeArrow } = this.props;

    const settings = {
      customPaging: i => (
        <button type="button" aria-label={`alt-viewSlide ${i + 1}`} />
      ),
      prevArrow: (
        <SlickPrevButton>
          <ChevronLeftOutlined
            aria-label="alt-prevArrowButton"
            className={classnames(
              customizeArrow.prevArrowClass,
              customizeArrow.arrowClass
            )}
            color={customizeArrow.color}
            fontSize={customizeArrow.fontSize}
          />
        </SlickPrevButton>
      ),
      nextArrow: (
        <SlickNextButton>
          <ChevronRightOutlined
            aria-label="alt-nextArrowButton"
            className={classnames(
              customizeArrow.nextArrowClass,
              customizeArrow.arrowClass
            )}
            color={customizeArrow.color}
            fontSize={customizeArrow.fontSize}
          />
        </SlickNextButton>
      ),
    };
    return (
      <Slider ref={this.setRef} {...settings} {...this.props}>
        {children}
      </Slider>
    );
  }
}

Carousel.propTypes = propTypes;
Carousel.defaultProps = defaultProps;

export default withStyles(carouselStyle)(Carousel);
