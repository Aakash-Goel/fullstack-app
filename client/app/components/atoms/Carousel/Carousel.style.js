const carouselStyles = () => ({
  '@global': {
    /* Slider */
    '.slick-slider': {
      position: 'relative',
      display: 'block',
      boxSizing: 'border-box',
      touchAction: 'pan-y',
      '-ms-touch-action': 'pan-y',
      '-webkit-tap-highlight-color': 'transparent',
    },
    '.slick-slider .slick-list, .slick-slider .slick-track': {
      transform: 'translate3d(0, 0, 0)',
    },
    '.slick-list': {
      position: 'relative',
      display: 'block',
      overflow: 'hidden',
      margin: 0,
      padding: 0,
    },
    '.slick-list:focus': {
      outline: 'none',
    },
    '.slick-list.dragging': {
      cursor: 'hand',
    },
    '.slick-track': {
      position: 'relative',
      top: 0,
      left: 0,
      display: 'block',
      marginLeft: 'auto',
      marginRight: 'auto',
    },
    '.slick-track::before, .slick-track::after': {
      display: 'table',
      content: '',
    },
    '.slick-track::after': {
      clear: 'both',
    },
    '.slick-loading .slick-track': {
      visibility: 'hidden',
    },
    '.slick-slide': {
      display: 'none',
      float: 'left',
      height: '100%',
      minHeight: '1px',
    },
    '[dir="rtl"] .slick-slide': {
      float: 'right',
    },
    '.slick-slide img': {
      display: 'block',
    },
    '.slick-slide.slick-loading img': {
      display: 'none',
    },
    '.slick-slide.slick-dragging img': {
      pointerEvents: 'none',
    },
    '.slick-initialized .slick-slide': {
      display: 'block',
    },
    '.slick-loading .slick-slide': {
      visibility: 'hidden',
    },
    '.slick-vertical .slick-slid': {
      display: 'block',
      height: 'auto',
      border: '1px solid transparent',
    },

    /* Arrows */
    '.slick-prev, .slick-next': {
      fontSize: 0,
      lineHeight: 0,
      position: 'absolute',
      top: '50%',
      display: 'block',
      padding: '0',
      cursor: 'pointer',
      border: 'none',
      outline: 'none',
      background: 'transparent',
      color: 'transparent',
      zIndex: 9,
      opacity: 0.75,
    },
    '.slick-prev:hover, .slick-prev:focus, .slick-next:hover, .slick-next:focus': {
      outline: 'none',
      opacity: 1,
    },
    '.slick-prev.slick-disabled, .slick-next.slick-disabled': {
      opacity: 0.25,
    },
    '.slick-prev': {
      left: '0px',
    },
    '[dir="rtl"] .slick-prev': {
      right: '0px',
    },
    '.slick-next': {
      right: '0px',
    },
    '[dir="rtl"] .slick-next': {
      right: 'auto',
      left: '0px',
    },
    '.slick-arrow.slick-hidden': {
      display: 'none',
    },

    /* Dots */
    '.slick-dots': {
      position: 'absolute',
      bottom: '.5em',
      display: 'block',
      width: '100%',
      padding: 0,
      margin: 0,
      listStyle: 'none',
      textAlign: 'center',
    },
    '.slick-dots li': {
      position: 'relative',
      display: 'inline-block',
      margin: '0 0.375em',
      padding: 0,
      cursor: 'pointer',
    },
    '.slick-dots li button': {
      display: 'block',
      cursor: 'pointer',
      width: '.5rem',
      height: '.5rem',
      padding: 0,
      lineHeight: 0,
      textIndent: '-999px',
      borderRadius: '50%',
      background: 'transparent',
      opacity: 0.5,
      border: `1px solid #999`,
    },
    '.slick-dots li button:hover, .slick-dots li button:focus': {
      opacity: 0.75,
      border: `1px solid #6D1B7B`,
    },
    '.slick-dots li.slick-active button': {
      opacity: 1,
      border: `1px solid #9c27b0`,
      background: `#9c27b0`,
    },
  },
});

export default carouselStyles;
