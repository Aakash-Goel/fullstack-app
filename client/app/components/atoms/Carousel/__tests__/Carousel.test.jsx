import React from 'react';
import TestRenderer from 'react-test-renderer';

import Carousel from '../Carousel';

describe('Carousel atom', () => {
  it('should renders correctly', () => {
    const carouselWrapper = TestRenderer.create(
      <Carousel>
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </Carousel>
    ).toJSON();
    expect(carouselWrapper).toMatchSnapshot();
  });
});
