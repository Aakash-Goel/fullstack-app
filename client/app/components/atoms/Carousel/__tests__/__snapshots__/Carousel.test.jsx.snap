// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel atom should renders correctly 1`] = `
<div
  className="slick-slider slick-initialized"
  dir="ltr"
>
  <div
    alt="alt-prevArrowButton"
    className="slick-arrow slick-prev"
    data-role="none"
    onClick={[Function]}
    style={
      Object {
        "display": "block",
      }
    }
    title="alt-prevArrowButton"
  >
    <svg
      aria-hidden="true"
      aria-label="alt-prevArrowButton"
      className="MuiSvgIcon-root-1 MuiSvgIcon-colorPrimary-2"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59z"
      />
    </svg>
  </div>
  <div
    className="slick-list"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={null}
    onMouseMove={null}
    onMouseUp={[Function]}
    onTouchCancel={null}
    onTouchEnd={[Function]}
    onTouchMove={null}
    onTouchStart={[Function]}
    style={Object {}}
  >
    <div
      className="slick-track"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      style={
        Object {
          "WebkitTransform": "translate3d(0px, 0px, 0px)",
          "WebkitTransition": "",
          "msTransform": "translateX(0px)",
          "opacity": 1,
          "transform": "translate3d(0px, 0px, 0px)",
          "transition": "",
        }
      }
    >
      <div
        aria-hidden={true}
        className="slick-slide slick-cloned"
        data-index={-1}
        onClick={[Function]}
        style={
          Object {
            "width": 0,
          }
        }
        tabIndex="-1"
      >
        <div>
          <div
            style={
              Object {
                "display": "inline-block",
                "width": "100%",
              }
            }
            tabIndex={-1}
          >
            3
          </div>
        </div>
      </div>
      <div
        aria-hidden={false}
        className="slick-slide slick-active slick-current"
        data-index={0}
        onClick={[Function]}
        style={
          Object {
            "outline": "none",
            "width": 0,
          }
        }
        tabIndex="-1"
      >
        <div>
          <div
            style={
              Object {
                "display": "inline-block",
                "width": "100%",
              }
            }
            tabIndex={-1}
          >
            1
          </div>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="slick-slide"
        data-index={1}
        onClick={[Function]}
        style={
          Object {
            "outline": "none",
            "width": 0,
          }
        }
        tabIndex="-1"
      >
        <div>
          <div
            style={
              Object {
                "display": "inline-block",
                "width": "100%",
              }
            }
            tabIndex={-1}
          >
            2
          </div>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="slick-slide"
        data-index={2}
        onClick={[Function]}
        style={
          Object {
            "outline": "none",
            "width": 0,
          }
        }
        tabIndex="-1"
      >
        <div>
          <div
            style={
              Object {
                "display": "inline-block",
                "width": "100%",
              }
            }
            tabIndex={-1}
          >
            3
          </div>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="slick-slide slick-cloned"
        data-index={3}
        onClick={[Function]}
        style={
          Object {
            "width": 0,
          }
        }
        tabIndex="-1"
      >
        <div>
          <div
            style={
              Object {
                "display": "inline-block",
                "width": "100%",
              }
            }
            tabIndex={-1}
          >
            1
          </div>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="slick-slide slick-cloned"
        data-index={4}
        onClick={[Function]}
        style={
          Object {
            "width": 0,
          }
        }
        tabIndex="-1"
      >
        <div>
          <div
            style={
              Object {
                "display": "inline-block",
                "width": "100%",
              }
            }
            tabIndex={-1}
          >
            2
          </div>
        </div>
      </div>
      <div
        aria-hidden={true}
        className="slick-slide slick-cloned"
        data-index={5}
        onClick={[Function]}
        style={
          Object {
            "width": 0,
          }
        }
        tabIndex="-1"
      >
        <div>
          <div
            style={
              Object {
                "display": "inline-block",
                "width": "100%",
              }
            }
            tabIndex={-1}
          >
            3
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    alt="alt-nextArrowButton"
    className="slick-arrow slick-next"
    data-role="none"
    onClick={[Function]}
    style={
      Object {
        "display": "block",
      }
    }
    title="alt-nextArrowButton"
  >
    <svg
      aria-hidden="true"
      aria-label="alt-nextArrowButton"
      className="MuiSvgIcon-root-1 MuiSvgIcon-colorPrimary-2"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6-6-6z"
      />
    </svg>
  </div>
</div>
`;
